cep-recipe-jinja.md

```diff
@@ -0,0 +1,527 @@
+<tr><td> Implementation </td><td>https://github.com/prefix-dev/rattler-build</td></tr>
+</table>
+
+## Abstract
+
+This CEP is part of the effort to strictly define a new recipe format. The previous CEPs are:
+
+- [CEP 13: A new recipe format](https://github.com/conda/ceps/blob/main/cep-13.md)
+- [CEP 14: A new recipe format: allowed keys and values](https://github.com/conda/ceps/blob/main/cep-14.md)
+
+Historically, `conda-build` recipes have relied on templating with `Jinja` for some
+"dynamic" functionality. For example, many recipes use the `version` of the
+package in multiple places (as package version, in the URL and the tests, for
+example). To make it easy to change recipes, Jinja has been used for some
+light-weight templating.
+
+The "old" recipe format has allowed arbitrary Jinja syntax (including set,
+if/else or for loops). The new recipe format only allows a subset of Jinja with
+the goal of always producing valid YAML files. In this CEP we clarify how Jinja
+is used in the new recipe format, what Jinja functions are available and how
+variables can be set and used.
+
+## Jinja in the new recipe format
+
+The new recipe format uses a subset of Jinja. Specifically, only "variable"
+expressions are allowed (no blocks such as `set`, `for` loops, `if`/`else`
+blocks, ...).
+
+A Jinja expression in the new recipe format looks like the following:
+
+`${{ version }}`
+
+Or if a function is involved:
+
+`${{ compiler('c') }}`
+
+Jinja expressions are also used in `if` statements and in the `skip` field of a
+recipe. In both instances, the `${{ ... }}` syntax is omitted.
+
+E.g.:
+
+```yaml
+build:
+  skip:
+    - osx  # This is a Jinja expression!
+
+requirements:
+  build:
+    - if: win and cuda  # This is a Jinja expression!
+      then:
+        - cudatoolkit
+```
+
+## Variables in the recipe
+
+The variables that are available in the Jinja context in the recipe come from
+two sources: the "variant configuration" file or the "context" section of the
+recipe.
+
+### The context section
+
+The context is a dictionary at the top-level of a recipe that maps keys to
+scalar values. The keys can be used by accessing them as variables in the Jinja
+expressions. For example:
+
+```yaml
+context:
+  version: "1.0.5"
+
+package:
+  version: ${{ version }}
+```
+
+Context evaluation must happen from top-to-bottom. That means a later value can
+reference an earlier one like so:
+
+```yaml
+context:
+  version: "1.0.5"
+  name_and_version: "pkg_${{ version | replace('.', '_') }}"  # evaluates to "pkg_1_0_5"
+```
+
+### Variables from variant configuration
+
+Any variable specified in the variant configuration file can be used in a recipe
+by using it in a Jinja expression. For example, with a variant config like:
+
+```yaml
+cuda:
+  - "have_cuda"
+  - "no_cuda"
+```
+
+This value can be used as follows, for example in an inline `if` expression:
+
+```yaml
+requirements:
+  host:
+    - ${{ "cudatoolkit" if cuda == "have_cuda" }}
```

Comment:
Should we add an explicit callout for the case like this where `expression` can evaluate into `""` or i guess `null` and that means this is excluded.

Reply:
We have something about that in line 500 - 510: https://github.com/conda/ceps/pull/71/files/c1028142a7f6e3e8a5d011667e0bec3b3f67ab3a#diff-0154c4169103373897853cfe5287bb7a1101289dd78c7fa30d2acba92b644938R500-R510